### ISSUE TITLE
Add public API baselines for rest of our implementation projects

### ISSUE THREAD

**Issue by dougbu** [dotnet/aspnetcore#24347]
If joining me in this endeavour, please assign yourself and add your name in parentheses on the line or lines below. Would rather not have multiple people working on adding the same files in parallel 😺 
- [x] src/Analyzers (@Pilchie) - contains no projects that need `PublicApi.*.txt` files ~~files exist, but aren't used by the build yet.~~
- [x] src/Antiforgery (@pilchie)
- [x] src/Azure (@pilchie) - needs a fix to the analyzer for Razor generator (dotnet/roslyn#47350)
- [x] src/Components (@dougbu)
- [x] src/Configuration.KeyPerFile (@pilchie)
- [x] src/DataProtection (@pilchie)
- [x] src/DefaultBuilder (@pilchie)
- [x] src/Features (@Pilchie)
- [x] src/FileProviders (@Pilchie)
- [x] src/Framework - contains no projects that need `PublicApi.*.txt` files (@Pilchie)
- [x] src/Grpc (@Pilchie)
- [x] src/HealthChecks (@Pilchie)
- [x] src/Hosting (@Pilchie)
- [x] src/Html (@dougbu)
- [x] src/Http (@dougbu)
- [x] src/Identity (@Pilchie) - needs a fix to the analyzer for Razor generator (dotnet/roslyn#47350)
- [x] src/Installers - contains no projects that need `PublicApi.*.txt` files
- [x] src/JSInterop (@Pilchie)
- [x] src/Localization (@Pilchie)
- [x] src/Logging.AzureAppServices (@Pilchie)
- [x] src/Middleware (@Pilchie)
- [x] src/ObjectPool (@Pilchie)
- [x] src/Razor (@Pilchie)
- [x] src/Security (@Pilchie)
- [x] src/Servers (@Pilchie)
- [x] src/SignalR (@Pilchie)
- [x] src/Testing (@Pilchie)
- [x] src/SiteExtensions (@Pilchie)
- [x] src/WebEncoders (@Pilchie)

I left MusicStore, ProjectTemplates, Shared, and Tools out because I'm pretty sure they don't contain `$(IsImplementationProject)` C# projects. If I'm wrong, please add to the list.

Mvc is already done.

**Comment by wtgodbe** [comment]
Seems like maybe the kind of thing I could do on ops duty? (Tuesday & Wednesday this week)

**Comment by dougbu** [comment]
The `PublicApi.*.txt` files aren’t really intended for manual intervention. To add them copy almost-empty files like src\Mvc\Mvc.Abstractions\src\PublicAPI.Shipped.txt into the directories you're adding baselines to e.g.
1. `cp .\src\Mvc\Mvc.Abstractions\src\PublicAPI.Shipped.txt .\src\Http\Http.Abstractions\src\PublicAPI.Shipped.txt`
2. `cp .\src\Mvc\Mvc.Abstractions\src\PublicAPI.Shipped.txt .\src\Http\Http.Abstractions\src\PublicAPI.Unshipped.txt`
3. `. .\activate.ps1`
4. `.\src\Http\startvs.cmd`
5. `F6` # or whatever your favourite build gesture is
6. Click on a RS0016 error
7. Right click in editor on underscored symbol or go straight to the “quick fix” icon to its left
![Fixer](https://user-images.githubusercontent.com/6431421/88596327-b654ec00-d019-11ea-8d42-e132ee36ab81.png). `Control-.` also works.
8. Choose `“Add Blah to public API”` / `“Fix all occurrences in … Solution”`
9. Click `Apply`
10. `F6` # yes, again to see if the fixer missed anything or if other RS00xx errors show up (not uncommon)
11. Suppress or fix other problems as needed but please suppress (if suppressing) using attributes and not globally or with `#pragma`s because attributes make the justification obvious
    - The fixer tends to crash VS or at least the current Int Preview version. Search for other attributes suppressing the error and copy it where you need it e.g. I suppressed RS0026 at https://github.com/dotnet/aspnetcore/blob/918d953767a8a9988c9e075254b70866e8b24e76/src/Mvc/Mvc.Core/src/Routing/ControllerLinkGeneratorExtensions.cs#L40
        - Add `using System.Diagnostics.CodeAnalysis;` if it's not already there.
    - Or, if the API is something new in 5.0, fix the problem as described in https://github.com/dotnet/roslyn/blob/master/docs/Adding%20Optional%20Parameters%20in%20Public%20API.md and related docs.
    - Reach out to me if you're not sure what action to take.

More information available in
- https://github.com/dotnet/roslyn-analyzers/blob/master/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md
- https://github.com/dotnet/roslyn-analyzers/blob/master/src/PublicApiAnalyzers/Microsoft.CodeAnalysis.PublicApiAnalyzers.md

**Comment by dougbu** [comment]
@wtgodbe if you don't see anything to improve in the project templates tests and you have time, sure take something here. But, it's pretty much the last ops-on-call priority from my perspective.

**Comment by Pilchie** [comment]
Working on a few of these today.

**Comment by dougbu** [comment]
Thanks @Pilchie❕

**Comment by dougbu** [comment]
One note I forgot: src\Mvc\Mvc.Abstractions\src\PublicAPI.Shipped.txt contains the single line
``` c#
#nullable enable
```
This may be incorrect in projects where #5680 hasn't been done. I'm not positive about this but recommend seeing if correct files are generated in such a project and, if necessary, starting with a completely empty file instead (where appropriate). Please let us know the results here😺

**Comment by dougbu** [comment]
> src/Analyzers (@Pilchie) - files exist, but aren't used by the build yet.

Why would we enable the files to be used for analyzers❔ Would be simple to change but I'm not sure it's the **Right Thing:tm:** to do.

**Comment by Pilchie** [comment]
Ok, `Components` is up for grabs, but the rest are done.  I can pick up Components if others want me to.

As for Analyzers, you're right.  It's *conceivable* that people could reference them to build their own analyzers, but I'm not worried about preventing that.

**Comment by dougbu** [comment]
I'll take Components. I started on it at one point because it was referenced in the `*.slnf` I used for HTML and HTTP at first. I'll finish that up very soon.

**Comment by dougbu** [comment]
> src/Analyzers (@Pilchie) - files exist, but aren't used by the build yet.

Should one of us remove those files❔

> src/Framework (may contain no projects that need PublicApi.*.txt files) (@Pilchie)
> src/Installers (may contain no projects that need PublicApi.*.txt files)

Confirmed src/Framework contains no C# code other than tests and src/Installers contains no `*.csproj` projects. Did you confirm the same thing @Pilchie❔

**Comment by Pilchie** [comment]
Yes and Yes.

**Comment by dougbu** [comment]
Note to self: File an issue to follow up on a dotnet/roslyn#47350 fix to enable the analyzers by default for all implementation project except our analyzers, tools, `msbuild` tasks, and similar.

**Comment by Pilchie** [comment]
I think this is done now @dougbu

**Comment by dougbu** [comment]
I agree and suggest you do the honours -- much of the work was your's❕ Many thanks

**Comment by Pilchie** [comment]
Fixed by a host of the linked PRs.
